File: app/jobs/business_enrichment_job.rb
class BusinessEnrichmentJob < ApplicationJob
  queue_as :default

  # Retry configuration for transient failures
  retry_on StandardError, wait: :exponentially_longer, attempts: 2 do |job, exception|
    contact_id = job.arguments.first
    contact = Contact.find_by(id: contact_id)
    Rails.logger.warn("Business enrichment failed for contact #{contact_id}: #{exception.message}")
  end

  # Don't retry on permanent failures
  discard_on ActiveRecord::RecordNotFound do |job, exception|
    Rails.logger.error("Contact not found for enrichment: #{exception.message}")
  end

  # Note: This job expects a contact_id (Integer), not a Contact object
  # This follows Sidekiq best practice of passing IDs to avoid serialization issues
  def perform(contact_id)
    contact = Contact.find(contact_id)
    # Skip if already enriched
    if contact.business_enriched?
      Rails.logger.info("Skipping contact #{contact.id}: already enriched")
      return
    end

    # Skip if not completed lookup yet
    unless contact.lookup_completed?
      Rails.logger.info("Skipping contact #{contact.id}: lookup not completed")
      return
    end

    # Check if business enrichment is enabled
    credentials = TwilioCredential.current
    unless credentials&.enable_business_enrichment
      Rails.logger.info("Business enrichment disabled in settings")
      return
    end

    # Perform enrichment
    Rails.logger.info("Enriching business data for contact #{contact.id}")

    success = BusinessEnrichmentService.enrich(contact)

    if success
      Rails.logger.info("Successfully enriched contact #{contact.id} with business data")
      # Note: Trust Hub and Email enrichment jobs are enqueued in parallel by EnrichmentCoordinatorJob
      # They will check business_enriched? before processing
    else
      Rails.logger.info("No business data found for contact #{contact.id}")
      # Don't mark as enriched if no data found - allows future re-attempts
      # Contact will be eligible for re-enrichment on next lookup
    end

  # Keep broad StandardError rescue at outermost level to ensure job failures are properly logged
  # and retried via Sidekiq's retry mechanism. This prevents silent failures in the background job queue.
  rescue StandardError => e
    Rails.logger.error("Unexpected error enriching contact #{contact.id}: #{e.class} - #{e.message}")
    Rails.logger.error(e.backtrace.join("\n"))
    raise
  end
end

File: laravel-app/app/Jobs/BusinessEnrichmentJob.php
<?php

namespace App\Jobs;

use App\Models\Contact;
use App\Models\TwilioCredential;
use App\Services\BusinessEnrichmentService;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use Exception;

class BusinessEnrichmentJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $contactId;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 2;

    /**
     * The number of seconds to wait before retrying.
     *
     * @var int
     */
    public $backoff = [10, 30];

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 60;

    /**
     * Create a new job instance.
     *
     * @param int $contactId
     */
    public function __construct(int $contactId)
    {
        $this->contactId = $contactId;
        $this->onQueue('default');
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        try {
            $contact = Contact::findOrFail($this->contactId);

            // Skip if already enriched
            if ($contact->business_enriched) {
                Log::info("Skipping contact {$contact->id}: already enriched");
                return;
            }

            // Skip if not completed lookup yet
            if (!$contact->lookup_completed) {
                Log::info("Skipping contact {$contact->id}: lookup not completed");
                return;
            }

            // Check if business enrichment is enabled
            $credentials = TwilioCredential::current();
            if (!$credentials || !$credentials->enable_business_enrichment) {
                Log::info("Business enrichment disabled in settings");
                return;
            }

            // Perform enrichment
            Log::info("Enriching business data for contact {$contact->id}");

            $service = new BusinessEnrichmentService();
            $success = $service->enrich($contact);

            if ($success) {
                Log::info("Successfully enriched contact {$contact->id} with business data");
            } else {
                Log::info("No business data found for contact {$contact->id}");
            }

        } catch (Exception $e) {
            Log::error("Unexpected error enriching contact {$this->contactId}: " . get_class($e) . " - {$e->getMessage()}");
            Log::error($e->getTraceAsString());
            throw $e;
        }
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::warning("Business enrichment failed for contact {$this->contactId}: {$exception->getMessage()}");
    }
}
